src/renderer/generic.zig

```diff
@@ -0,0 +1,2873 @@
+                // so that we can cleanly deinit our GPU state.
+                for (0..buf_count) |_| self.frame_sema.wait();
+                for (&self.frames) |*frame| frame.deinit();
+            }
+
+            /// Get the next frame state to draw to. This will wait on the
+            /// semaphore to ensure that the frame is available. This must
+            /// always be paired with a call to releaseFrame.
+            pub fn nextFrame(self: *SwapChain) error{Defunct}!*FrameState {
+                if (self.defunct) return error.Defunct;
+
+                self.frame_sema.wait();
+                errdefer self.frame_sema.post();
+                self.frame_index = (self.frame_index + 1) % buf_count;
+                return &self.frames[self.frame_index];
+            }
+
+            /// This should be called when the frame has completed drawing.
+            pub fn releaseFrame(self: *SwapChain) void {
+                self.frame_sema.post();
+            }
+        };
+
+        /// State we need duplicated for every frame. Any state that could be
+        /// in a data race between the GPU and CPU while a frame is being drawn
+        /// should be in this struct.
+        ///
+        /// While a draw is in-process, we "lock" the state (via a semaphore)
+        /// and prevent the CPU from updating the state until our graphics API
+        /// reports that the frame is complete.
+        ///
+        /// This is used to implement double/triple buffering.
+        const FrameState = struct {
+            uniforms: UniformBuffer,
+            cells: CellTextBuffer,
+            cells_bg: CellBgBuffer,
+
+            grayscale: Texture,
+            grayscale_modified: usize = 0,
+            color: Texture,
+            color_modified: usize = 0,
+
+            target: Target,
+            /// See property of same name on Renderer for explanation.
+            target_config_modified: usize = 0,
+
+            /// Custom shader state, this is null if we have no custom shaders.
+            custom_shader_state: ?CustomShaderState = null,
+
+            /// A buffer containing the uniform data.
+            const UniformBuffer = Buffer(shaderpkg.Uniforms);
+            const CellBgBuffer = Buffer(shaderpkg.CellBg);
+            const CellTextBuffer = Buffer(shaderpkg.CellText);
+
+            pub fn init(api: GraphicsAPI, custom_shaders: bool) !FrameState {
+                // Uniform buffer contains exactly 1 uniform struct. The
+                // uniform data will be undefined so this must be set before
+                // a frame is drawn.
+                var uniforms = try UniformBuffer.init(api.uniformBufferOptions(), 1);
+                errdefer uniforms.deinit();
+
+                // Create the buffers for our vertex data. The preallocation size
+                // is likely too small but our first frame update will resize it.
+                var cells = try CellTextBuffer.init(api.fgBufferOptions(), 10 * 10);
```

Comment:
If this is likely too small and we're going to resize anyways, is there any reason to allocate 100 elements to begin with? Why not just do `1`. _Really_, we should avoid the allocation completely since it just forces a double-allocation on every new path anyways but the cost of that is probably ignorable.

Not a big deal, but just seems funny to even put 100 here.

Reply:
I believe I copied this from the existing code, I agree that we may as well start with 1 like with the 1x1 textures

Reply:
Reduced size to 1, also reduced the initial atlas texture sizes to 1 since they'll be resized anyway